processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -262,13 +262,20 @@ else if ( !method.isUpdateMethod() &&
         }
 
         private MethodReference getFinalizerMethod(Type resultType) {
-            if ( method.getReturnType().isVoid() ||
-                resultType.getEffectiveType().isAssignableTo( resultType ) ) {
+            Type returnType = method.getReturnType();
+            if ( returnType.isVoid()) {
                 return null;
             }
-            BuilderType builderType = resultType.getBuilderType();
+            Type mappingType = method.isUpdateMethod() ? resultType : resultType.getEffectiveType();
```

Comment:
I tried to parse this code.. but I remain having problems understanding what this code does.. IIUC if this is an update method we can have: 1) either a builder (in which we should call the finaliser ) or 2) not a builder (and then we should not call the finaliser ).. 

The method implies that it returns a finaliser.. However that is not what it does. Proposal: split into 2 methods that can be used on the call site.. 

Proposal:
1. `getFinalizerMethod` returns a finalizer when it can.
2. `useFinalizer` or `callFinalizer` (boolean) takes all scenarios into account.